src/snowflake/connector/connection.py

```diff
@@ -913,100 +925,111 @@ def __open_connection(self):
 
         # Setup authenticator
         auth = Auth(self.rest)
-        if self.auth_class is not None:
-            if type(
-                self.auth_class
-            ) not in FIRST_PARTY_AUTHENTICATORS and not issubclass(
-                type(self.auth_class), AuthByKeyPair
-            ):
-                raise TypeError("auth_class must be a child class of AuthByKeyPair")
-                # TODO: add telemetry for custom auth
-            self.auth_class = self.auth_class
-        elif self._authenticator == DEFAULT_AUTHENTICATOR:
-            self.auth_class = AuthByDefault(
-                password=self._password,
-                timeout=self._login_timeout,
-                backoff_generator=self._backoff_generator,
-            )
-        elif self._authenticator == EXTERNAL_BROWSER_AUTHENTICATOR:
-            self._session_parameters[PARAMETER_CLIENT_STORE_TEMPORARY_CREDENTIAL] = (
-                self._client_store_temporary_credential if IS_LINUX else True
-            )
-            auth.read_temporary_credentials(
-                self.host,
-                self.user,
-                self._session_parameters,
+
+        if self._session_token and self._master_token:
+            auth._rest.update_tokens(
+                self._session_token,
+                self._master_token,
+                self._master_validity_in_seconds,
             )
-            # Depending on whether self._rest.id_token is available we do different
-            #  auth_instance
-            if self._rest.id_token is None:
-                self.auth_class = AuthByWebBrowser(
-                    application=self.application,
-                    protocol=self._protocol,
-                    host=self.host,
-                    port=self.port,
-                    timeout=self._login_timeout,
-                    backoff_generator=self._backoff_generator,
-                )
-            else:
-                self.auth_class = AuthByIdToken(
-                    id_token=self._rest.id_token,
-                    application=self.application,
-                    protocol=self._protocol,
-                    host=self.host,
-                    port=self.port,
+            auth._rest._heartbeat()
```

Comment:
I'm curious what kind of error will be raised if invalid/expired token is provided?

Reply:
This is what it shows:
```snowflake.connector.errors.InterfaceError: 250003 (08001): 401 Unauthorized: post https://ig86763.preprod12.us-west-2.aws.snowflakecomputing.com:443/session/heartbeat?requestId=4a63f07e-4928-4f98-b02d-58a4e7151f73&request_guid=148cac21-a4da-4341-bb24-862f8a6211c7```

I think I can try to write a test for that.

Reply:
yeah let's add a test for it -- error behavior is also part of the public apis

Reply:
I added a test for it and initially I edited errors.py to add another error type (the Unauthorized one) but I noticed it caused issues with the old driver test due to imports, so I instead went with ProgrammingError since technically speaking the tokens aren't valid anymore. What's your take on this?

Reply:
I don't think we need to introduce another error type, whatever is returned nowadays shall be good

Reply:
Okay, roger that.